### PR TITLE
Bootstrap documentation

### PR DESCRIPTION
**Pull request by tealeg**
Partial fix for #17 - documentation for the `Consumer` will follow when it has been added.

### CODE REVIEW

File: consumer/doc.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// This package will eventually contain the Consumer type.  The
+// Consumer is felice's primary entrance point for recieving messages
```
Comment (asdine) [inline consumer/doc.go]:
```suggestion
// Consumer is felice's primary entrance point for receiving messages
```

File: consumer/handler/doc.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// The handler package defines the mechanisms by which a recieved
```
Comment (asdine) [inline consumer/handler/doc.go]:
```suggestion
// The handler package defines the mechanisms by which a received
```

File: consumer/handler/doc.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+//
+//    type MyFooHandler struct { }
+//
+//    func (mfh MyFooHandler) HandleMessage(msg *message.Message) error {
+//        fmt.Printf("%+v", *msg)
+//    }
+//
+// This approach has the advantage of not actually requiring you to
+// import the handler package when defining handlers for use with the
+// conusmer.Consumer.
```
Comment (asdine) [inline consumer/handler/doc.go]:
```suggestion
// consumer.Consumer.
```

File: consumer/handler/doc.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+//
+// The second approach is to cast a function to the HandlerFunc type
+// defined in this package:
+//
+//    h := handler.HandlerFunc(func(msg *message.Message) error {
+//        fmt.Printf("%+v", *msg)
+//    })
+//
+// Handlers can also be managed in Collections.  The Collection struct
```
Comment (asdine) [inline consumer/handler/doc.go]:
```suggestion
// Handlers can also be managed in Collections. The Collection struct
```

File: consumer/handler/doc.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+// will allow exactly one handler to be associated with each topic,
+// and will, on demand return a list of all topics for which a handler
+// has been registered.  For example:
```
Comment (asdine) [inline consumer/handler/doc.go]:
```suggestion
// has been registered. For example:
```

File: consumer/handler/doc.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+//
+// The Handler interface defines the signature for all felice
+// Handlers.  There are two common ways to comply with this interface.
+// The first is simply to create a type with the HandleMessage
+// function:
+//
+//    type MyFooHandler struct { }
+//
+//    func (mfh MyFooHandler) HandleMessage(msg *message.Message) error {
+//        fmt.Printf("%+v", *msg)
```
Comment (asdine) [inline consumer/handler/doc.go]: :tulip: I think we also need to explain what returning an error implies, about the fact that when a user returns an error, the message is replayed after a delay, indefinitely.
But maybe after the PR with the consumer?

File: consumer/handler/doc.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+//
+// Handlers can also be managed in Collections.  The Collection struct
+// will allow exactly one handler to be associated with each topic,
+// and will, on demand return a list of all topics for which a handler
+// has been registered.  For example:
+//
+//    col := handler.Collection{}
+//
+//    col.Set("my-topic", handler.HandlerFunc(func(msg *message.Message)) {
+//        fmt.Printf("Got message on my-topic: %+v", *msg)
```
Comment (asdine) [inline consumer/handler/doc.go]: :warning: The signature doesn't match, the handler func must return an error

File: consumer/handler/interface.go [hunk 8]
```diff
@@ -2,7 +2,10 @@ package handler
 
 import "github.com/heetch/felice/message"
 
-// A Handler handles a Message.
+// Handler is the interface for handling consumed messages.  You can
```
Comment (asdine) [inline consumer/handler/interface.go]:
```suggestion
// Handler is the interface for handling consumed messages. You can
```

File: consumer/handler/interface.go [hunk 9]
```diff
@@ -2,7 +2,10 @@ package handler
...
+// either add a compliant HandleMessage function to some type by hand,
+// or make use of the HandlerFunc.  Note that felice Handlers recieve
```
Comment (asdine) [inline consumer/handler/interface.go]:
```suggestion
// or make use of the HandlerFunc. Note that felice Handlers receive
```

File: message/doc.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+//
+// You can create a new Message by calling New:
+//
+//    msg := New("my-topic", "simple string message")
+//
+// The value passed as the 2nd argument can be any Go type that can be
+// marshaled with encoding/json.
+//
+// New can also be passed zero, one or many additional Options.  An
+// Option is a function that recieves a pointer to the Message and can
```
Comment (asdine) [inline message/doc.go]:
```suggestion
// Option is a function that receives a pointer to the Message and can
```

File: message/message.go [hunk 11]
```diff
@@ -9,17 +9,23 @@ import (
 	uuid "github.com/satori/go.uuid"
 )
 
-// Message contains informations about the message to be sent to Kafka.
+// Message represents a message to be sent via Kafka, or recieved from
```
Comment (asdine) [inline message/message.go]:
```suggestion
// Message represents a message to be sent via Kafka, or received from
```

File: message/message.go [hunk 12]
```diff
@@ -9,17 +9,23 @@ import (
...
+// it.  When using Felice's Consumer, any Handlers that you register
+// will recieve Messages as they're arguments.  When using the Felice
+// Producer, you will be sending Messages.  When making a Message to
+// be sent it essential that you use the New function to do so.
```
Comment (asdine) [inline message/message.go]:
```suggestion
// be sent it is essential that you use the New function to do so.
```

File: consumer/handler/doc.go (hunk 6 above)
Comment (YassineE) [reply consumer/handler/doc.go]: > But maybe after the PR with the consumer?

🌷 I think it can be added to this PR.

File: message/message.go [hunk 13]
```diff
@@ -9,17 +9,23 @@ import (
 	uuid "github.com/satori/go.uuid"
 )
 
-// Message contains informations about the message to be sent to Kafka.
+// Message represents a message to be sent via Kafka, or recieved from
+// it.  When using Felice's Consumer, any Handlers that you register
+// will recieve Messages as they're arguments.  When using the Felice
```
Comment (YassineE) [inline message/message.go]:
```suggestion
// will receive Messages as they're arguments.  When using the Felice
```

File: message/doc.go [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+// modify it directly prior to it being returned by New.  Two
+// predefined Options exists in felice: Header and Key.
+//
+// For example, if you want to create a Message with a custom header
+// field you could request it as follows:
+//
+//    msg := New("my-topic", "cold potatoes ain't hot!", Header("subject", "potatoes"))
+//
+package message
```
Comment (YassineE) [inline message/doc.go]: 🌷 Should we add details about the default headers like the `Message-Id` and the `Produced-At`?

File: message/option.go [hunk 15]
```diff
@@ -1,16 +1,26 @@
 package message
 
-// Option is used to customize a message before sending it.
+// Option is a function type that recieves a pointer to a Message and
```
Comment (YassineE) [inline message/option.go]:
```suggestion
// Option is a function type that receives a pointer to a Message and
```

File: message/option.go [hunk 16]
```diff
@@ -1,16 +1,26 @@
...
+// modifies it in place.  Options are intended to customize a message
+// before sending it.  You can do this either by passing them as
+// paramters to the New function, or by calling them directly against
```
Comment (YassineE) [inline message/option.go]:
```suggestion
// parameters to the New function, or by calling them directly against
```

File: message/option.go [hunk 17]
```diff
@@ -1,16 +1,26 @@
...
+// past to New will be the value that appears on the Message.
 func Header(k, v string) Option {
 	return func(m *Message) {
 		m.Headers[k] = v
 	}
 }
 
-// Key option specifies a key for the message.
+// Key is an Option that specifies a key for the message.  You should
+// only pass ths once to the New function, but if you pass it multiple
```
Comment (YassineE) [inline message/option.go]:
```suggestion
// only pass this once to the New function, but if you pass it multiple
```

File: consumer/handler/doc.go (hunk 6 above)
Comment (tealeg) [reply consumer/handler/doc.go]: I've documented this on the interface's definition of `HandleMessage`.

File: consumer/handler/doc.go (hunk 7 above)
Comment (tealeg) [reply consumer/handler/doc.go]: Good catch!

**Pull request merged**